### PR TITLE
[IMP] mail: markdown in discuss channels

### PR DESCRIPTION
**Pull request by phenix-factory**

### CODE REVIEW

File: addons/mail/models/mail_link_preview.py [hunk 1]
```diff
@@ -32,6 +33,10 @@ def _create_from_message_and_notify(self, message):
         if tools.is_html_empty(message.body):
             return self
         urls = OrderedSet(html.fromstring(message.body).xpath('//a[not(@data-oe-model)]/@href'))
+        markdown_link_re = r"\[.+?\]\(\s*([^)]+)\s*\)"
+        matchs = re.findall(markdown_link_re, message.body)
+        for url in matchs:
+            urls.add(url)
```
Comment (tsm-odoo) [inline addons/mail/models/mail_link_preview.py:39]: maybe?
```suggestion
        markdown_link_re = r"\[.+?\]\(\s*([^)]+)\s*\)"
        urls.extend(re.findall(markdown_link_re, message.body))
```

File: addons/mail/static/src/core/common/message_model.js [hunk 2]
```diff
@@ -40,6 +49,23 @@ export class Message extends Record {
     attachments = Record.many("Attachment", { inverse: "message" });
     author = Record.one("Persona");
     body = Record.attr("", { html: true });
+    bodyMarkdown = Record.attr("", {
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/message_model.js:52]: Confusing field name, this is not containing `bodyMarkdown` but `bodyHtml`/`bodyMarkup` or something like that.

File: addons/mail/static/src/core/common/message_model.js [hunk 3]
```diff
@@ -40,6 +49,23 @@ export class Message extends Record {
...
+        compute() {
+            if (!this.isBodyEmpty) {
+                if (!this.store.markdownLoaded) {
+                    loadMarkdownAssets.loadMarkdown().then(() => this.store.markdownLoaded = true);
+                    return this.body;
+                }
+                const htmlDoc = parser.parseFromString(this.body, "text/html");
+                const markdownElements = htmlDoc.querySelectorAll("odoo-markdown");
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/message_model.js:60]: This could be checked before, if there's no `odoo-markdown` we don't need to load the lib.

File: addons/mail/static/src/core/common/store_service.js [hunk 4]
```diff
@@ -148,6 +148,7 @@ export class Store extends BaseStore {
             description: _t("Notify everyone"),
         },
     ];
+    markdownLoaded = false;
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/store_service.js:151]: It's strange to have a loaded state in a different file than the load method, they should be together.

File: addons/mail/static/src/core/common/thread_model.js [hunk 5]
```diff
@@ -1046,17 +1046,20 @@ export class Thread extends Record {
...
+            let prettyBody = await prettifyMessageContent(
                 body,
                 this.store.getMentionsFromText(body, {
                     mentionedChannels,
                     mentionedPartners,
                 })
             );
+            if (this.model === "discuss.channel") {
+                prettyBody = `<odoo-markdown>${prettyBody}</odoo-markdown>`;
+            }
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/thread_model.js:1058]: I'm confused why do we add `<odoo-markdown>` in several different places. It's done in `getMessagePostParams` which is called here above.

File: addons/mail/static/src/core/common/message_model.js [hunk 6]
```diff
@@ -346,12 +372,16 @@ export class Message extends Record {
             mentionedChannels,
             mentionedPartners,
         });
+        let prettyBody = await prettifyMessageContent(body, validMentions);
+        if (this.thread.model === "discuss.channel") {
+            prettyBody = `<odoo-markdown>${prettyBody}</odoo-markdown>`;
+        }
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/message_model.js:378]: This is probably not a good way of doing it.

There should be no processing of any input at all inside markdown code blocks in particular.

**Review by seb-odoo — COMMENTED** [review]
Preliminary check

File: addons/im_livechat/controllers/main.py [hunk 7]
```diff
@@ -59,6 +59,13 @@ def get_emoji_bundle(self):
         stream = request.env['ir.binary']._get_stream_from(asset.js())
         return stream.get_response()
 
+    @http.route("/im_livechat/markdown_bundle", type="http", auth="public", cors="*")
+    def get_markdown_bundle(self):
+        bundle = "mail.assets_markdown"
```
[the remark this answers is not in the document]
Comment (phenix-factory) [reply addons/im_livechat/controllers/main.py]: This is a copy/past from  `get_emoji_bundle`.
I moved the logic inside a `stream_bundle` function, it's easier that way.

[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/im_livechat/controllers/main.py]: I see but it's unnecessary variables in that method too, better not copy that.

File: addons/mail/static/src/core/common/store_service.js (hunk 4 above)
Comment (phenix-factory) [reply addons/mail/static/src/core/common/store_service.js:151]: I understand, but this library should not be loaded for each message, so it's marked loaded once for all models.

File: addons/mail/static/src/core/common/message_model.js (hunk 2 above)
Comment (phenix-factory) [reply addons/mail/static/src/core/common/message_model.js]: I think `bodyHtml` is fine

File: addons/mail/static/src/core/common/store_service.js (hunk 4 above)
Comment (seb-odoo) [reply addons/mail/static/src/core/common/store_service.js:151]: The load function should be in store not message

File: addons/mail/models/mail_link_preview.py [hunk 8]
```diff
@@ -32,6 +33,8 @@ def _create_from_message_and_notify(self, message):
         if tools.is_html_empty(message.body):
             return self
         urls = OrderedSet(html.fromstring(message.body).xpath('//a[not(@data-oe-model)]/@href'))
+        markdown_link_re = r"\[.+?\]\(\s*([^)]+)\s*\)"
+        urls.update(re.findall(markdown_link_re, message.body))
```
[the remark this answers is not in the document]
Comment (phenix-factory) [reply addons/mail/models/mail_link_preview.py:37]: > Don't we want to parse markdown links only inside markdown message?

It doesn't matter, even if I write a markdown link inside a non-markdown message, the URL will be extracted and the OrderedSet will not allow the same URL twice. The result is the same.

File: addons/mail/static/src/core/common/store_service.js (hunk 4 above)
Comment (phenix-factory) [reply addons/mail/static/src/core/common/store_service.js:153]: I moved everything inside a store function

File: addons/im_livechat/static/src/embed/common/message_model_patch.js [hunk 9]
```diff
@@ -1,6 +1,5 @@
 import { Message } from "@mail/core/common/message_model";
 import { Record } from "@mail/core/common/record";
-
```
Comment (seb-odoo) [inline addons/im_livechat/static/src/embed/common/message_model_patch.js:3]: unnecessary diff in unchanged file

File: addons/im_livechat/static/src/embed/common/store_service_patch.js [hunk 10]
```diff
@@ -3,6 +3,9 @@ import { SESSION_STATE } from "@im_livechat/embed/common/livechat_service";
 import { Store, storeService } from "@mail/core/common/store_service";
 
 import { patch } from "@web/core/utils/patch";
+import { url } from "@web/core/utils/urls";
+import { loadJS } from "@web/core/assets";
```
Comment (seb-odoo) [inline addons/im_livechat/static/src/embed/common/store_service_patch.js:7]: (minor) alphabetical assets before utils

File: addons/im_livechat/static/tests/embed/livechat_service.test.js [hunk 11]
```diff
@@ -122,7 +122,7 @@ test("Only necessary requests are made when creating a new chat", async () => {
                 temporary_id: 0.81,
             },
             post_data: {
-                body: "Hello!",
+                body: "<odoo-markdown>Hello!</odoo-markdown>",
```
Comment (seb-odoo) [inline addons/im_livechat/static/tests/embed/livechat_service.test.js:125]: AL suggested to use `<pre>` tag with `data-` to allow this new format to be fully "compatible" with HTML in case the lib is not available, so that it just renders as a markdown code block in standard HTML.

File: addons/mail/models/mail_link_preview.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (seb-odoo) [reply addons/mail/models/mail_link_preview.py:37]: Why should we look for links inside markdown if the result is the same?

Also note: this should exclude links from inside code blocks.

File: addons/mail/static/src/core/common/message.scss [hunk 12]
```diff
@@ -44,16 +44,21 @@
 .o-mail-Message-body {
     font-family: "text-emoji", $font-family-base;
 
-    & > p {
+    p:last-child {
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/message.scss:47]: image
lists generate unwanted extra margin at the end too

File: addons/mail/static/src/core/common/message.scss [hunk 13]
```diff
@@ -106,3 +111,124 @@
...
+        color: #b31d28;
+        background-color: #ffeef0
+    }
+    .hljs-char.escape_,
+    .hljs-link,
+    .hljs-params,
+    .hljs-property,
+    .hljs-punctuation,
+    .hljs-tag {
+        /* purposely ignored */
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/message.scss:231]: why is that purpose? and what's the point of defining an empty block if we don't care?

File: addons/mail/static/src/core/common/message.scss [hunk 14]
```diff
@@ -106,3 +111,124 @@
...
+
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/message.scss:232]: unnecessary empty line

File: addons/mail/static/src/core/common/message.xml [hunk 15]
```diff
@@ -88,7 +88,7 @@
                                                         <em t-if="message.subject and !message.isSubjectSimilarToThreadName and !message.isSubjectDefault" class="mb-1 me-2">Subject: <t t-out="props.messageSearch?.highlight(message.subject) ?? message.subject"/></em>
                                                         <div t-if="message.message_type and message.message_type.includes('email')" t-ref="shadowBody"/>
                                                         <t t-elif="state.showTranslation" t-out="message.translationValue"/>
-                                                        <t t-elif="message.body" t-out="props.messageSearch?.highlight(message.body) ?? message.body"/>
+                                                        <t t-elif="message.body" t-out="props.messageSearch?.highlight(message.bodyHtml) ?? message.bodyHtml"/>
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/message.xml:91]: condition elif should be on bodyHtml too probably

File: addons/mail/static/src/core/common/message_model.js [hunk 16]
```diff
@@ -12,8 +12,13 @@ import { _t } from "@web/core/l10n/translation";
 import { user } from "@web/core/user";
 import { omit } from "@web/core/utils/objects";
 import { url } from "@web/core/utils/urls";
+import { markdown } from "@mail/utils/common/markdown";
+import { markup } from "@odoo/owl";
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/message_model.js:16]: (minor) order alphabetical and by module

File: addons/mail/static/src/core/common/message_model.js [hunk 17]
```diff
@@ -12,8 +12,13 @@ import { _t } from "@web/core/l10n/translation";
...
 
 const { DateTime } = luxon;
+
+const parser = new DOMParser();
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/message_model.js:20]: What's the difference between instantiating a new one for each usage and having a global one?

File: addons/mail/static/src/core/common/message_model.js [hunk 18]
```diff
@@ -40,6 +44,23 @@ export class Message extends Record {
...
     author = Record.one("Persona");
     body = Record.attr("", { html: true });
+    bodyHtml = Record.attr("", {
+        compute() {
+            if (!this.isBodyEmpty) {
+                const htmlDoc = parser.parseFromString(this.body, "text/html");
+                const markdownElements = htmlDoc.querySelectorAll("odoo-markdown");
+                if (!this.store.markdownLoaded && markdownElements.length > 0) {
+                    this.store.loadMarkdownBundle();
+                    return this.body;
```
Comment (seb-odoo) [inline addons/mail/static/src/core/common/message_model.js:54]: To be improved, at least empty is be better than flicker of content.
To consider: not rendering message at all while this is not done.

File: addons/web/static/tests/_framework/module_set.hoot.js [hunk 19]
```diff
@@ -513,6 +513,9 @@ const WHITE_LISTED_KEYS = [
     "Chart", // Chart.js
     "L", // Leaflet
     "lamejs", // LameJS
+    "hljs", // highlight.js
+    "marked", // marked.js
+    "markedHighlight", // markedHighlight.js
```
Comment (seb-odoo) [inline addons/web/static/tests/_framework/module_set.hoot.js:518]: list looked alphabetical

File: addons/im_livechat/static/tests/embed/livechat_service.test.js (hunk 11 above)
Comment (phenix-factory) [reply addons/im_livechat/static/tests/embed/livechat_service.test.js:125]: I see, I would argue the following:

* In HTML5 we can create whatever markup we want, it doesn't make it incompatible with anything I know of.
* For the people with markdown working, it will be invalid HTML since `<pre>` doesn't accept a lot of elements as children:
image
When using a custom element: 
image

File: addons/mail/static/src/core/common/message.scss (hunk 13 above)
Comment (phenix-factory) [reply addons/mail/static/src/core/common/message.scss:231]: I do not know, this is a copy of the GitHub CSS theme provided by the highlight.js lib. I did not change it.
I guess we could clean it.

File: addons/mail/static/src/core/common/message.scss (hunk 14 above)
Comment (phenix-factory) [reply addons/mail/static/src/core/common/message.scss:232]: I wonder if it's really unnecessary? Maybe this is a way to "empty" the rule?

File: addons/mail/static/src/core/common/message_model.js (hunk 17 above)
Comment (phenix-factory) [reply addons/mail/static/src/core/common/message_model.js:20]: Instantiate it once instead of two? I guess it's not that much of an optimization.
Maybe it could be a static of message class?

File: addons/mail/static/src/core/common/message_model.js (hunk 18 above)
Comment (phenix-factory) [reply addons/mail/static/src/core/common/message_model.js:54]: > To consider: not rendering message at all while this is not done.

I initially wanted to do this, but @alexkuhn doesn't want to delay the rendering.
I still think markdown should be loaded earlier, when discuss is "started" when a channel is open, or something like that.

Comment (alexkuhn) [reply addons/mail/static/src/core/common/message_model.js:55]: @phenix-factory I'm not against delaying the rendering, but it depends on how you implement it. If this is a `onWillStart(() => loadMarkdownBundle())` then this is not good, because this blocks parented rendering too.
If this is a `t-if` in `<Thread>` then why not.

We could fine-tune the moment we load markdown, but this is not a motivation to load it so early that we don't have to deal with async loading while we want to show a message. "When discuss is started" is at page load for setup of services and models and this is not ok because we want lazy-load. At channel opening is not much better, because a message is not necessarily shown there (messaging menu item, chat bubble preview, notif toaster, etc.). At least being in the use of `message.body` that contains some markdown ensures exact moment we need it and works in all current (and future) showing of message body in code.

Comment (alexkuhn) [reply addons/mail/static/src/core/common/message_model.js:55]: And not rendering the message is fine, but functionally speaking we don't really care. Sure this is theoretically less work for OWL, but this doesn't matter in terms of any perceivable performance gains.

Comment (phenix-factory) [reply addons/mail/static/src/core/common/message_model.js:55]: I'm experimenting with web workers, maybe this will become a nonissue.

Comment (seb-odoo) [reply addons/mail/static/src/core/common/message_model.js:55]: This is quite complex, I was going to post the following message (I still keep it below, because it is good food for thought) but then I realized it's quite confusing to display a channel (if it's empty) and then no longer display it while markdown is loading when a message is posted or if it receives a message.

And it's even more complicated with inbox.

Isn't it core enough that we shouldn't do lazy loading?
If we have discuss we almost certainly have channels (general), if we have channels we more than likely have messages, so either way we need the lib all the time.

Unless we only load it when a message is really displayed, which is the only way to lazy load it in a way that makes sense, but then we have all the issues I mentioned to resolve (flicker, combination with other features, etc).

---

I think it should be a condition _before_ rendering thread component, so in parents: discuss, public page, chat window. Messaging menu is making exception as it only displays the "preview", so it can be handled differently.

If the there is at least one markdown message inside the thread, it should wait on markdown being loaded before displaying the component.

We absolutely can't render empty thread/messages or thread with non-parsed markdown because it will mess up with load older, scroll restore, and stuff like that, due to them being heavily based on timings and rendering. We could refactor them to take delayed loading but it's probably more work?

The "trigger" for when loading the lib could be based on model state like it is now, it doesn't need to wait on the message being actually displayed. Knowledge in store of a message with markdown is more than enough to assume this message will soon enough be rendered.

Comment (seb-odoo) [reply addons/mail/static/src/core/common/message_model.js:55]: If most of the size is for code formatting, we should immediately load the parsing lib all the time, and only lazy-load code syntax when a specific language is used.
